### PR TITLE
fix: launch DynamoDB local instance for development automatically

### PR DESCRIPTION
**Pull request by chejimmy**
# Description

Using the [`OnApplicationBootstrap` lifecycle hook](https://docs.nestjs.com/fundamentals/lifecycle-events) to launch a DynamoDB local instance for development environment.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] core e2e passing
- [x] lint passing

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Legal

This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    await ddbLocalLaunch(this.dbConfig.port);
+    this.logger.log('DynamoDB local instance launched');
+    await this.createApiResourceTable();
+  }
+
+  private async createApiResourceTable() {
+    const ddbClient = new DynamoDBClient({ endpoint: this.dbConfig.endpoint });
+    let attempt = 0;
+
+    do {
```
Comment (NorbertNader) [inline apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts:43]: Why do we need to retry table creation?

**Review by NorbertNader — APPROVED** [review]

File: apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+const TABLE_CREATION_ATTEMPT_NUMBER = 30;
+const TABLE_CREATION_ATTEMPT_DELAY = 1000; // in milliseconds
+
+@Injectable()
+export class DynamoDbLocalSetupService implements OnApplicationBootstrap {
+  private readonly logger = new Logger(DynamoDbLocalSetupService.name);
+
+  constructor(
+    @Inject(databaseConfig.KEY)
+    private dbConfig: ConfigType<typeof databaseConfig>,
```
Comment (tracy-french) [inline apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts:29]: neat

File: apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  Logger,
+  OnApplicationBootstrap,
+} from '@nestjs/common';
+import { ConfigType } from '@nestjs/config';
+import { launch as ddbLocalLaunch } from 'dynamodb-local';
+import { databaseConfig } from '../config/database.config';
+
+const apiResourceTable =
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  require('../../api-resource-table-properties.js') as CreateTableCommandInput;
```
Comment (tracy-french) [inline apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts:19]: Not a blocker: Can the table properties be in TypeScript instead?

**Review by tracy-french — APPROVED** [review]

File: apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts (hunk 1 above)
Comment (chejimmy) [reply apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts:43]: It is waiting for the local instance to spin up. The logic could be separated into [1] wait for spin up, and then [2] table creation request when we have more requests.

File: apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts (hunk 3 above)
Comment (chejimmy) [reply apps/core/src/lifecycle-hooks/dynamodb-local-setup.ts:19]: I want to keep the `api-resource-table-properties.js` common between `jest-dynamodb-config.js` and `dynamodb-local-setup.ts`. I can look into if that's possible.

**Pull request merged**